### PR TITLE
Remove microdata in html flow and replace by JSON-LD for classic theme

### LINKED ISSUE #22867: Move structured data to a separate file
#### Is your feature request related to a problem?

Currently, all microdatas are managed in their respective templates of the markup they aim to enrich, this leads to several things :
- It's harder to maintain, because there are a lot of files to inspect in order to change these
- It's harder to override from a template, because the template needs to override tons of tpl files, and it leads to maintainability problems because when we update the classic theme, you need to update several templates only because of microdatas
- It's not respecting seperation of concern as you've a mix between "Shown markup" and "SEO Microdatas" which make templates harder to read and understand

#### Describe the solution you'd like

Having one file, containing every microdatas of the classic theme, that you could override on your template in order to personnalize it way more, fix theme-sided changes.

Example :
https://github.com/prestarocket-agence/classic-rocket/blob/master/templates/_partials/microdata-jsonld.tpl

### PR DESCRIPTION
**Pull request by fdonnet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Try to propose a solution for all the recurent issues with Google Rich Text validator. (REOPEN => I made a silly git mistake I was not able to roll back)
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #22867
| How to test?      | Fresh version of Presta, go on Goolge rich text validator, test index, category and product pages => clean
| Possible impacts? | Change in classic template. Add two new files included in head.tpl

The goal of this PR is to open the discussion about the microdata management.
A lot of issues are open about this topic recently.
The proposition is to keep the meta tags in the html flow but remove the microdata information and create a specific file that centralizes all the microdata and use JSON-LD for the declaration.
_inspired a lot by github =>prestarocket-agence/classic-rocket_

**themes/classic/templates/_partials/microdata-jsonld.tpl**  :

This file is included in the head.tpl.
For the moment it's not perfect but it centralizes the info for futher improvements.

It takes care of this kind of microdata
- Webpage
- Organization
- Search link
- Category pages (ListItem but that only ref the product pages no product details in the ListItem)
- Products (need to be standardized)
- ~~Product combinations declared as an offer for the product (with stock and correct price) => great stuff ;)~~ Not implemented for standard shop (people will be able to override)
- Breadcrumb

The product JSON declaration is not perfect. But like explained it's to open the discussion about the "centralization" of the concept and after that, maybe people who go deep;) will create new smarty variables that will help to enrich the microdata declaration.

~~people find comibnations (colors, flavors) too. => solve the problem when you have single cannonical url per product and your combinations will be skipped by Google bot.~~ (available in the first commit of this PR, but removed to remain standard)

**Other files => remove each instance of itemprop and itemscope**
**Trick for pagination pages**

Hope it will help...

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23042)

### CODE REVIEW

**Comment by fdonnet** [comment]
That was a very bad idea to remove my install-dev folder and push to my repo... My Old pull request was #23151 closed by a reset I tried to do... sry for that (noob inside).

**Comment by fdonnet** [comment]
@boubkerbribri 
@nesrineabdmouleh 

On the PR #23151 

- I installed PHP 7.4
- No error with my PR on the product page related to the extraction you sent to me yesterday... so I think it's not linked to my php7.4 ui test failure...
- I installed the shop and I run the sanity test locally (with php 7.4 enable) and I have this error related to the selector main h1 (timeout) on product.tpl
- And after that, I don't know why but I deleted my install-dev folder I pushed to my remote origin... (morning mess)
- Big panic, I reseted and the PR was closed... so I pls find a new one here with my changes :) :) => first PR guy

Sry for that.
If someone can help (when the tests will be finished) to identitfy the issue with php7.4 UI test it can be super cool, because I cannot identity the problem on my side
THX

this selector in UI Test 7.4 this.productName = '#main h1'; seems to not work...

**Comment by boubkerbribri** [comment]
Hello @fdonnet ,
I'm sorry but me and @nesrineabdmouleh can't help you fix this error because it's not related to the tests, maybe @PrestaShop/prestashop-maintainers 

What I can assure you is that I checkout you PR and the error exists on php7.4
image

**Comment by fdonnet** [comment]
@boubkerbribri Hi, sry for the delay, lot of other things to go after... 
I searched a little bit but I cannot reproduce on my side on WSL and mod php7.4... It's a common issue with php 7.4 (it seems) => like you can see php is not one of my language at all ;). And I didn't find a way to identify and correct the mistake...

EDIT : Can confirm, php 7.4 (only my PR) => no error on product page... hard to debug.
EDIT2 : Try to add an isset on property checking without any conviction (cannot reproduce on my side)... I don't know if I kept the correct meaning... and I don't understand why sometimes we can access smarty with "." array and sometimes with "->"... Like I said c# boy here ;)

**Comment by fdonnet** [comment]
@Hlavtox if you can point me a way to have the combination url (landing page) in smarty var or by any other way, I think I have a very sexy solution for the combinations offers ;) . On my side I just add the combination id in the base url but it is very dirty.

**Comment by fdonnet** [comment]
UI PHP 7.4 failed again... if someone can help me to identify the thing I need to change. I think I have an empty data and a insufficent test on it and in PHP7.4 it put a false or something like that (not with precedent vs of php). I m suffering ;)

**Comment by NeOMakinG** [comment]
I can't get why it would fail on 7.4 and not on previous versions, this is pretty strange

**Comment by NeOMakinG** [comment]
Running tests again as the fail isn't from the PR

File: themes/classic/templates/catalog/_partials/product-prices.tpl
Comment (Progi1984) [inline themes/classic/templates/catalog/_partials/product-prices.tpl]: That shouldn't be removed ?

**Review by Progi1984 — COMMENTED** [review]
A small feedback

File: themes/classic/templates/catalog/_partials/product-prices.tpl
Comment (fdonnet) [reply themes/classic/templates/catalog/_partials/product-prices.tpl:40]: It's already present in the JSON LD declaration, do you think it's necessary to keep it in the meta ? If all other itemprop etc are removed do you think this meta with only conntent=currency is really understandable by the robots ? I readd it no problem if needed because I removed that part without special knowledge on the topic ;)

File: themes/classic/templates/catalog/_partials/product-prices.tpl
Comment (Progi1984) [inline themes/classic/templates/catalog/_partials/product-prices.tpl:41]:
```suggestion
          <span>
```

**Review by Progi1984 — COMMENTED** [review]
A last one

**Review by Progi1984 — DISMISSED** [review]

**Comment by fdonnet** [comment]
@Progi1984 if you have way to get product combination url i will be able to add combinations support in product-jsonld.tpl.

Something like that =>
Get product info for the cannonical page (in product if no combination, in default combination if product with combination) and declare other combinations as additionnal offers.

I will make a new PR for that... maybe....

File: themes/classic/templates/catalog/_partials/product-cover-thumbnails.tpl
Comment (NeOMakinG) [inline themes/classic/templates/catalog/_partials/product-cover-thumbnails.tpl]:
```suggestion
              itemprop="image"
```

File: themes/classic/templates/catalog/_partials/product-images-modal.tpl
Comment (NeOMakinG) [inline themes/classic/templates/catalog/_partials/product-images-modal.tpl]:
```suggestion
                      itemprop="image"
                      height="148"
```

File: themes/classic/templates/catalog/_partials/product-cover-thumbnails.tpl
Comment (fdonnet) [reply themes/classic/templates/catalog/_partials/product-cover-thumbnails.tpl]: done

File: themes/classic/templates/catalog/_partials/product-images-modal.tpl
Comment (fdonnet) [reply themes/classic/templates/catalog/_partials/product-images-modal.tpl]: done

**Comment by fdonnet** [comment]
Hi think I have a UI test issue @nesrineabdmouleh. 

line 30 in tests/UI/pages/FO/product/index.js
`this.productPrice = ${this.productPricesBlock} .current-price span;`

This span has no content in my PR because I removed all microdata and I don't know if I need to roll back for this particular one.
If I remember well it was a suggestion during the review to remove it.

line 40 In my PR themes/classic/templates/catalog/_partials/product-prices.tpl
` <div class="current-price">
          <span>
           ...`

In develop branch => ` <div class="current-price">
          <span itemprop="price" content="{$product.rounded_display_price}">`

The tests seem to fail because of that... maybe you can help @NeOMakinG too (need I re-add this content attribute in the span html tag)? Find another selector seems not so easy ?
 
EDIT : when I said re-add, I think only the content attr not the itemprop...

File: themes/classic/templates/catalog/_partials/product-images-modal.tpl
Comment (fdonnet) [reply themes/classic/templates/catalog/_partials/product-images-modal.tpl]: without the itemprop (I removed all in my PR => image links are declared in JSONLD)

**Comment by fdonnet** [comment]
trying to push with this content attribute in span in themes/classic/templates/catalog/_partials/product-prices.tpl (L40)

Only to pass the UI tests or to see If I had more side effects on other UI selectors... (microdata test selectors...)

**Comment by fdonnet** [comment]
Ok, I m back to the eternal error I have on PHP7.4 (only) UI tests :):) good news... but xD

**Comment by Progi1984** [comment]
Ping @boubkerbribri for Sanity Tests : May be could you help @fdonnet  ?

**Comment by Progi1984** [comment]
@fdonnet It seems that the same bug as there : https://github.com/PrestaShop/PrestaShop/pull/23151#issuecomment-775034582

**Comment by fdonnet** [comment]
> 
> 
> @fdonnet It seems that the same bug as there : [#23151 (comment)](https://github.com/PrestaShop/PrestaShop/pull/23151#issuecomment-775034582)

I was not able to reproduce on my side... for me it's linked to the selector h1 (i don't know why but in 7.4 it doesn't work)... @Progi1984 .
@NeOMakinG helped in the past =>
`I can't get why it would fail on 7.4 and not on previous versions, this is pretty strange`
`Running tests again as the fail isn't from the PR` ... so i don't know how to correct

**Comment by NeOMakinG** [comment]
I'm gonna try to take a look on your branch later

**Comment by fdonnet** [comment]
@NeOMakinG it seems that I found the issue.
It's linked to when you preview a product without any cover image. 
I tested like this : `{if isset($product.cover)}` and it raised an error in 7.4 when it was empty so I replaced by `{if !empty($product.cover)}`... I think it will correct the problem... and it was really a bug on my side ;)

File: tests/UI/pages/FO/product/index.js
Comment (Progi1984) [inline tests/UI/pages/FO/product/index.js]:
```suggestion
  constructor() {
```

Comment (fdonnet) [reply tests/UI/pages/FO/product/index.js]: Oula the keyboard miss touch sry ;) Corrected.

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
@fdonnet The problem is always here in 7.4

**Comment by fdonnet** [comment]
@Progi1984, yep same issue when a product is defined with combinations and no combination images and no base cover image is defined... correction has been push now.

**Comment by fdonnet** [comment]
@Progi1984 I re-pushed without the combinations part (it was removed before) because I cannot retrieve the product combination url on the product page (not able to propose JSON LD offers correctly). Don't know why this commit was removed in my rebase => git pro). But now I hope, I will be able te propose a first (not complete step) for this JSONLD PR ;)

**Comment by fdonnet** [comment]
@NeOMakinG in the future, if we will be able to add the combination product url in the $combinations smarty var like @Hlavtox made it for ean13 in #23055. I will be able to add all combinations as offers for a product. The benefit is that all the potential products will be exposed on the cannonical page and it will be great.

**Comment by fdonnet** [comment]
@Progi1984 new failed on 7.3 UI this time :) selector : .growl-message:last-of-type.... This time it s maybe not my fault, who knows

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
@fdonnet Fixed (with restarting CI)

**Review by NeOMakinG — DISMISSED** [review]
Great :)

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (Hlavtox) [inline themes/classic/templates/_partials/microdata/product-jsonld.tpl]: @fdonnet I think you should use InStock only if the product is really in stock - `"availability":{if $product.quantity > 0} "http://schema.org/InStock"{else} "http://schema.org/OutOfStock"{/if},`

See: https://support.google.com/merchants/answer/6324448#zippy=%2Cproduct-inventory-ran-out%2Cproduct-is-available-for-sale%2Cvariants

Comment (fdonnet) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl]: hi @Hlavtox, if i remember well you asked me to add $product.allow_oosp > 0 in my old PR... so really, it's like you prefer. Do I remove it and keep only the product quantity check ?

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/microdata/product-jsonld.tpl:36]: I guess `$shop.name` always exists? 🤔 shall we always treat `$shop.name` as brand? 🤔 I'm not sure

File: themes/classic/templates/_partials/microdata/product-list-jsonld.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/microdata/product-list-jsonld.tpl]:
```suggestion
        {foreach from=$listing.products item=item key="position" name=productsForJsonLd}
```

File: themes/classic/templates/_partials/microdata/product-list-jsonld.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/microdata/product-list-jsonld.tpl]:
```suggestion
            }{if !$smarty.foreach.productsForJsonLd.last},{/if}
```

File: themes/classic/templates/_partials/pagination-seo.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/pagination-seo.tpl]:
```suggestion
```

File: themes/classic/templates/_partials/pagination-seo.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/pagination-seo.tpl]:
```suggestion
    {$page_nb = 1}
```

File: themes/classic/templates/_partials/pagination-seo.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/pagination-seo.tpl]:
```suggestion
        {$page_nb = $smarty.get.page|intval|default:1}
```

Comment (kpodemski) [reply themes/classic/templates/_partials/pagination-seo.tpl]: You can use `default:1` and omit checking for isset 🤔

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (Hlavtox) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl]: @fdonnet Maybe, having second thoughts. I leave it up to you, your PR. 🤣 

@jf-viguier Wdyt? Should we put in stock also for backorder products? It would attract more customers, but I know google complains that my xml product data were different than the real data on the page.

File: themes/classic/templates/_partials/head.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/head.tpl]: this should definitely be in block, what about third party modules like `Advanced Search` or `Amazzing filters`? I think it would not be compatible

File: themes/classic/templates/_partials/pagination-seo.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/pagination-seo.tpl]:
```suggestion
    {$prev = false}
```

I found this kind of variable definition more readable

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (fdonnet) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl:36]: It's really a fallback, if no manufacturer is defined... it seems that people hate to have Google Rich text warnings.. it's to avoid this situation and propose a clean but not perfect rich text test result (limit issues opened about this topic)... the people will be able to override the template if needed. 
We only need to be aligned with "what is a base version" => "less warning but not perfect" or "not complete, with more warnings, but no tweak" ;)

File: themes/classic/templates/_partials/head.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/head.tpl]: I'd put it into

`{block name='header_microdata'}`

**Review by kpodemski — COMMENTED** [review]
Some feedback but overall very good job 💪

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (kpodemski) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl:36]: @fdonnet ok, you're right 👍

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (Hlavtox) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl]: @fdonnet Aaaahaaa, originally it is done like this: `<link itemprop="availability" href="{$product.seo_availability}"/>`. I think we should keep it, wdyt? So `"availability": "{$product.seo_availability}",`

Comment (kpodemski) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl]: @Hlavtox from what I know Google is planning to stop searching for differences between site and xml 🤔

Comment (Hlavtox) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl]: @kpodemski Hmm, nice point. Anyway, I think we should stick to what is provided by the core.

I proposed an issue with the availablity itself here: https://github.com/PrestaShop/PrestaShop/issues/23534

File: themes/classic/templates/_partials/head.tpl
Comment (fdonnet) [reply themes/classic/templates/_partials/head.tpl]: `  {block name='head_pagination_seo'}
    {include file="_partials/pagination-seo.tpl"}
  {/block}`

File: themes/classic/templates/_partials/head.tpl
Comment (fdonnet) [reply themes/classic/templates/_partials/head.tpl]: {block name='head_microdata'} yep, to keep consistency with other head block in head.tpl (done)

**Comment by fdonnet** [comment]
@kpodemski thx a lot for the review. All propositions integrated, question open about $shop.name in brand in case of no manufacturer.

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (Hlavtox) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl:36]: @fdonnet Shop name as manufacturer is a very good idea, if it's missing!!!

File: themes/classic/templates/_partials/microdata/product-jsonld.tpl
Comment (fdonnet) [reply themes/classic/templates/_partials/microdata/product-jsonld.tpl]: @Hlavtox corrected. Put the responsability of the value in the defined core smarty var ;) thx

**Review by Progi1984 — DISMISSED** [review]

File: themes/classic/templates/_partials/head.tpl
Comment (kpodemski) [inline themes/classic/templates/_partials/head.tpl]: I didn't catch it earlier but it would be cool to have it override possible as well, could you add block here @fdonnet ?

maybe 

`{block name='head_open_graph'}`

**Review by kpodemski — CHANGES_REQUESTED** [review]
One more little change and it's 🆗 for me

@NeOMakinG could you look at it as well?

File: themes/classic/templates/_partials/head.tpl
Comment (fdonnet) [reply themes/classic/templates/_partials/head.tpl]: yes, perfect, done.

**Review by kpodemski — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by NeOMakinG** [comment]
Hello @fdonnet,

Thanks a lot for the amazing work.

I did the QA on your PR and it seems that it contain an issue on every pages (url is not fine), see this report :

https://search.google.com/test/rich-results?id=S8qph4hGW_mzIahlF25P7g

Also tested on category page, it contains errors but I've the feeling that it's due to my environment, so google can't reach the URL, it shouldn't be an issue.

https://search.google.com/test/rich-results?id=seVMLmb2UwnIjGz48yN2bA

Product page :

https://search.google.com/test/rich-results?id=BNv3roZndl_bLU4H9uKLHQ

I think except the Logos error, we can merge it when it's fixed :)

**Comment by fdonnet** [comment]
hello @NeOMakinG : google validator raise an error if you test on your localhost (non valid url). You need to test it from another url, or change the url in the code you upload if you used the "code" method.

File: themes/classic/templates/_partials/microdata/head-jsonld.tpl
Comment (NeOMakinG) [inline themes/classic/templates/_partials/microdata/head-jsonld.tpl]: This is causing the error

Comment (fdonnet) [reply themes/classic/templates/_partials/microdata/head-jsonld.tpl]: Yes thx, I saw that is not only linked to localhost (sometimes it is not accepted for validation in the Google tool). Apparently, shop.logo smarty var contains the full url path now... so I made the modification. Thx a lot for the detection.

**Review by NeOMakinG — APPROVED** [review]
I tested with the fix:
https://search.google.com/test/rich-results?id=d-gSZPtkd3p4T5GjTKImNg

It's fine now (errors are generated because my urls aren't reachable by Google)

**Comment by NeOMakinG** [comment]
As it's a QA by dev, feel free to merge it when it gets 2 approves!

Good job @fdonnet

**Comment by fdonnet** [comment]
Thx, @NeOMakinG for the reviews, the advice and the support !

**Comment by NeOMakinG** [comment]
Trying to run tests again

**Review by Progi1984 — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @fdonnet & @NeOMakinG for QA ;)

**Comment by Hlavtox** [comment]
@fdonnet You're the boss!!!

**Comment by fdonnet** [comment]
@Hlavtox thx for your help too on polishing a lot of things ! Cool to be merged ;)

**Comment by Progi1984** [comment]
@fdonnet @Hlavtox Good job guys :muscle:

File: themes/classic/templates/_partials/microdata/head-jsonld.tpl
Comment (mgielecinski) [inline themes/classic/templates/_partials/microdata/head-jsonld.tpl:60]: hello, this is bad
{$urls.shop_domain_url}{$shop.logo} --> {$shop.logo}

Comment (fdonnet) [reply themes/classic/templates/_partials/microdata/head-jsonld.tpl:60]: @mgielecinski already modified in d40c7e2a905d28cbad0ecc85ed4b4d2531c45361

Comment (mgielecinski) [reply themes/classic/templates/_partials/microdata/head-jsonld.tpl:60]: in organization yes, but in line 60 still is error

Comment (fdonnet) [reply themes/classic/templates/_partials/microdata/head-jsonld.tpl:60]: @mgielecinski oh yeah I see :) Thx a lot. @Progi1984 for info I opened #24239 (pull request) to correct this based on mgielecinski laser eyes recommendation.

**Pull request merged**
